---
title: "scRNAseq with Seurat"
author: "Holly A. R. Giles"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
      code_folding: "hide" 
---

------------------------------------------------------------------------

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

```{r loadLibraries}

library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)

```

# Setup the Seurat Object

In this vignette, we are analysing a dataset of .... available at .... There are [number of cells] single cells that were sequenced on the [Illumina NextSeq 500]. The raw data can be found [here](insert%20link).

We start by reading in the data:

1.  `Read10X()` returns a UMI count matrix.

    *The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).*

2.  `CreateSeuratObject` uses the count matrix to create a `Seurat` object.

    *The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset.*

A note on `Seurat` objects:

Full description [here](https://github.com/satijalab/seurat/wiki). `Seurat` objects contain an `Assay` class, with slots for raw counts (\@counts slot), normalized data (\@data slot), and scaled data for dimensional reduction (\@scale.data slot), and `DimReduc` class. You can access and add data using `GetAssayData`, `SetAssayData`, `[['']]` and `$`.

Note: You may get warnings about suppressed column names and feature names cannot have underscores ('\_'), replacing with dashes ('-').

```{r loadData}

#continue from here, find where data is 
# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "/Users/HollyGiles/Documents/Projects/seurat/data/pbmc3k/filtered_gene_bc_matrices/hg19/")

# Initialise the Seurat object with the raw (non-normalised data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)

pbmc

```

```{r}
# Lets examine a few genes in the first 30 cells
pbmc.data[c("CD3D","TCL1A","MS4A1"), 1:30]
```

The `.` values in the matrix represent 0s (no molecules detected). Since most values in an scRNA-seq matrix are 0, Seurat uses a sparse-matrix representation whenever possible. This results in significant memory and speed savings for Drop-seq/inDrop/10x data. 

# Pre-processing workflow

The steps below encompass the pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalisation and scaling, and the detection of highly variable features.

## QC and selecting cells for further analysis

Here we explore QC metrics and filter cells based on defined criteria.

The QC metrics are as follows:

-   The number of unique genes detected in each cell.
    -   Low-quality cells or empty droplets will often have very few genes
    -   Cell doublets or multiplets may exhibit an aberrantly high gene count
-   Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
-   The percentage of reads that map to the mitochondrial genome
    -   Low-quality / dying cells often exhibit extensive mitochondrial contamination

### Unique genes and total molecules

Check the metadata object to get the number of unique genes and total molecules (calculated during `CreateSeuratObject()`).

```{r QC, fig.height=7, fig.width=13}
    # Show QC metrics for the first 5 cells
    head(pbmc@meta.data, 5)
```

### Mitochondrial reads

Calculate mitochondrial QC metrics with the `PercentageFeatureSet()` function, which calculates the percentage of counts originating from a set of features

Use the set of all genes starting with `MT-` as a set of mitochondrial genes

```{r mitochondrialReads, fig.height=7, fig.width=13}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```

###    Visualise QC metrics and filter cells

-   Filter cells that have unique feature counts over 2,500 or less than 200
-   Filter cells that have \>5% mitochondrial counts

```{r qc2, fig.height=7, fig.width=13}

#Visualise QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualise feature-feature relationships, but can be used for anything calculated by the object, 
# i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt") 
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
plot1 + plot2


#filter based on QC metrics
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

------------------------------------------------------------------------

# Data normalisation

We employ a global-scaling normalisation method "LogNormalize" that normalises the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalised values are stored in `pbmc[["RNA"]]@data`.

```{r normaliseData}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 1e4)
```

# Identification of highly variable features (feature selection)

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Focusing on these genes in downstream analysis helps to highlight **biological signal in single-cell datasets**.

The procedure in Seurat is described in detail [here](https://doi.org/10.1016/j.cell.2019.05.031), and is implemented in the `FindVariableFeatures()` function. By default, the function returns 2,000 features per dataset. These will be used in downstream analysis, like PCA.

```{r var_features, fig.height=5, fig.width=11}
pbmc <- FindVariableFeatures(pbmc, selection.method = 'vst', nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```

------------------------------------------------------------------------

# Scaling the data

## Linear Transformation

Next, we apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The `ScaleData()` function:

-   Shifts the expression of each gene, so that the mean expression across cells is 0
-   Scales the expression of each gene, so that the variance across cells is 1
    -   This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
-   The results of this are stored in `pbmc[["RNA"]]@scale.data`

Scale for all genes:

```{r regress, fig.height=7, fig.width=11, results='hide'}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

Faster option: scale only for previously identified variable features

(Scaling is an essential step in the Seurat workflow, but only on genes that will be used as input to PCA. NB PCA and clustering results are unaffected by only scaling variable genes. However, Seurat heatmaps require genes to be scaled, to make sure highly-expressed genes don't dominate the heatmap).

```{r regressvar, fig.height=7, fig.width=11, results='hide',eval = FALSE}
pbmc <- ScaleData(pbmc)
```

## Regress out unwanted sources of variation

Use the `ScaleData()` function to remove unwanted sources of variation from a single-cell dataset and 'regress out' heterogeneity associated with e.g. cell cycle stage, or mitochondrial contamination.

```{r regressvarmt, fig.height=7, fig.width=11, results='hide',eval = FALSE}
pbmc <- ScaleData(pbmc, vars.to.regress = 'percent.mt')
```

In case I need to run more sophisticated normalisation:

For advanced users who would like to use this functionality, we strongly recommend the use of our new normalisation workflow, `SCTransform()`. The method is described in this [paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1), with a separate vignette using Seurat v3 [here](sctransform_vignette.html).

 

------------------------------------------------------------------------

# Perform linear dimensional reduction

## Run PCA on the scaled data.

By default, only the previously determined variable features are used as input, but can be defined using `features` argument.

```{r pca,results='hide'}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```

## Visualise the PCs

Seurat provides several useful ways of visualising both cells and features that define the PCA, including `VizDimReduction()`, `DimPlot()`, and `DimHeatmap()`

```{r pca_viz, message=TRUE}
# Examine and visualize PCA results a few different ways
print(pbmc[['pca']], dims = 1:5, nfeatures = 5)
VizDimLoadings(pbmc, dims = 1:2, reduction = 'pca')
DimPlot(pbmc, reduction = 'pca')
```

## PCA Heatmap

`DimHeatmap()` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. Plotting a subset of cells by setting `cells` to a number plots the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though this a supervised analysis, it remains a valuable tool for exploring correlated feature sets.

Visualise PC1

```{r single-heatmap}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```

Visualise PC1 - PC15

```{r multi-heatmap, fig.height=15, fig.width=9}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```

# Determine the dimensionality of the dataset

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset.

Three approaches to consider when selecting the number of PCs:

1.  Supervised: exploring PCs to determine relevant sources of heterogeneity (as above). This could be used in conjunction with GSEA.

2.  Implement a statistical test based on a random null model (a little more time-consuming for large datasets, and may not return a clear PC cutoff).

3.  Heuristic method with `ElbowPlot` . Commonly used, and can be calculated instantly.

    In the example here, all three approaches yielded similar results, and you could be justified in choosing anything between PC 7-12 as a cutoff.

Method 2:

[Macosko *et al*](http://www.cell.com/abstract/S0092-8674(15)00549-8), implements a resampling test inspired by the JackStraw procedure. This method randomly permutes a subset of the data (1% by default) and rerun PCA, constructing a 'null distribution' of feature scores, and repeat this procedure. This identifies 'significant' PCs as those who have a strong enrichment of low p-value features.

```{r jackstraw, fig.height=6, fig.width=10}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More approximate techniques such as those implemented in ElbowPlot() can be used to reduce computation time
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
```

The `JackStrawPlot()` function provides a visualisation tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). 'Significant' PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a sharp drop-off in significance after the first 10-12 PCs.

```{r jsplots, fig.height=6, fig.width=10}
JackStrawPlot(pbmc, dims = 1:15)
```

Method 3:

An alternative heuristic method generates an 'Elbow plot': a ranking of principle components based on the percentage of variance explained by each one (`ElbowPlot()` function). In this example, we observe an 'elbow' around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs.

```{r elbow_plot, fig.height=6, fig.width=10}
ElbowPlot(pbmc)
```

Factors to consider, in addition to the above:

-   Repeat downstream analyses with a different number of PCs (10, 15, or even 50!) and the results often do not differ dramatically.
-   Err on the higher side when choosing this parameter. For example, performing downstream analyses with only 5 PCs does significantly and adversely affect results.

------------------------------------------------------------------------

# Cluster the cells

Seurat v3 applies a graph-based clustering approach, building upon initial strategies in ([Macosko *et al*](http://www.cell.com/abstract/S0092-8674(15)00549-8)). Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partitioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [[SNN-Cliq, Xu and Su, Bioinformatics, 2015]](http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract) and CyTOF data [[PhenoGraph, Levine *et al*., Cell, 2015]](http://www.ncbi.nlm.nih.gov/pubmed/26095251). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'.

As in PhenoGraph, a KNN graph is first constructed based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the `FindNeighbors()` function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, modularity optimisation techniques are next applied (such as the Louvain algorithm (default) or SLM [[SLM, Blondel *et al*., Journal of Statistical Mechanics]](http://dx.doi.org/10.1088/1742-5468/2008/10/P10008)), to iteratively group cells together, with the goal of optimizing the standard modularity function. The `FindClusters()` function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. Setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the `Idents()` function.

```{r cluster, fig.height=5, fig.width=7}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)

# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)
```

------------------------------------------------------------------------

# Run non-linear dimensional reduction (UMAP/tSNE)

The goal of tSNE and UMAP algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. NB Cells within the graph-based clusters determined above should co-localise on these dimension reduction plots. As input to the UMAP and tSNE, use the same PCs as input to the clustering analysis.

```{r tsne, fig.height=5, fig.width=7}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages = "umap-learn")
pbmc <- RunUMAP(pbmc, dims = 1:10)
```

```{r tsneplot, fig.height=5, fig.width=7}
# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
DimPlot(pbmc, reduction = 'umap')
```

------------------------------------------------------------------------

# Finding differentially expressed features (cluster biomarkers)

### Identify genes that are cluster biomarkers

Here we find markers that define each of the clusters via differential expression. `FindAllMarkers()` identifies positive and negative markers of a single cluster (specified by `ident.1)`, compared to all other cells. You can also test groups of clusters vs. each other, or against all cells. `FindMarkers()` is for more specific comparisons.

Arguments:

The `min.pct` argument requires a feature to be detected at a minimum percentage in either of the two groups of cells

The `thresh.test` argument requires a feature to be differentially expressed (on average) by some amount between the two groups.

Its possible to set both to 0, but with a dramatic increase in time since this will test a large number of features that are unlikely to be highly discriminatory.

To speed up these computations, `max.cells.per.ident` can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significant and the most highly differentially expressed features will likely still rise to the top.

```{r markers1, fig.height=8, fig.width=15}
# find all markers of cluster 2
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
head(cluster2.markers, n = 5)

# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)

# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#show the top two maerks in each cluster
pbmc.markers %>% group_by(cluster) %>% slice_max(n = 2, order_by = avg_log2FC)
```

### Using other test parameters for differential expression

Seurat has several tests for differential expression which can be set with the test.use parameter (see our [DE vignette](de_vignette.html) for details). For example, the ROC test returns the 'classification power' for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r markersroc, fig.height=8, fig.width=15}
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```

### Visualising biomarker expression

There are several tools for visualising marker expression:

1.  `VlnPlot()` shows expression probability distributions across clusters

2.  `FeaturePlot()` visualises feature expression on a tSNE or PCA plot

    Also:

3.  `RidgePlot()`

4.  `CellScatter()`

5.  `DotPlot()`

### Vln plot

```{r Vlnplots, fig.height=10, fig.width=15}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = 'counts', log = TRUE)
```

### Feature plot

```{r Featureplots, fig.height=10, fig.width=15}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"))
```

### Heatmap

`DoHeatmap()` generates an expression heatmap for given cells and features. Here we plot the top 20 markers (or all markers if less than 20) for each cluster.

```{r clusterHeatmap, fig.height=8, fig.width=15}
pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

------------------------------------------------------------------------

# Assigning cell type identity to clusters

Use canonical markers to easily match the unbiased clustering to known cell types.

### Cell type definitions

| Cluster ID | Markers       | Cell Type    |
|------------|---------------|--------------|
| 0          | IL7R, CCR7    | Naive CD4+ T |
| 1          | CD14, LYZ     | CD14+ Mono   |
| 2          | IL7R, S100A4  | Memory CD4+  |
| 3          | MS4A1         | B            |
| 4          | CD8A          | CD8+ T       |
| 5          | FCGR3A, MS4A7 | FCGR3A+ Mono |
| 6          | GNLY, NKG7    | NK           |
| 7          | FCER1A, CST3  | DC           |
| 8          | PPBP          | Platelet     |

### Labelled plot

```{r labelplot, fig.height=5, fig.width=9}
#define cluster IDs
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono", "NK", "DC", "Platelet")

names(new.cluster.ids) <- levels(pbmc)

pbmc <- RenameIdents(pbmc, new.cluster.ids)

DimPlot(pbmc, reduction = 'umap', label = TRUE, pt.size = 0.5) + NoLegend()
```

# Save objects

```{r save.img, include=FALSE}

plot <- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + 
  guides(colour = guide_legend(override.aes = list(size = 10)))

#ggsave(filename = "../inst/images/UMAP.jpg", height = 7, width = 12, plot = plot, quality = 50)
```

```{r saveobject, eval=FALSE}
#saveRDS(pbmc, file = "../data/seuratObject.rds")
```

<details>

<summary>**Session Info**</summary>

```{r}
sessionInfo()
```

</details>
