---
title: "scRNAseq data read in from GEO"
author: "Holly A. R. Giles"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
      code_folding: "hide" 
---

set this up as a template
the try for this data 

then carry on with rest of analysis 
------------------------------------------------------------------------

In this vignette, we read in a dataset of .... available at .... There are [number of cells] single cells that were sequenced on the [Illumina NextSeq 500]. The raw data can be found [here](insertlink).

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)

args <- commandArgs(trailingOnly = TRUE)

```

```{r loadLibraries, include=FALSE}

#library(tidyverse)
library(Seurat)
library(patchwork)
library(Matrix)
library(GEOquery)

```

```{r plotDir, include=FALSE}

plotDir = ifelse(exists(".standalone"), "", "../../inst/figs/") 
if(plotDir!="") if(!file.exists(plotDir)) dir.create(plotDir)

```

```{r defineAesthetics, include=FALSE}

source("/Users/HollyGiles/Documents/R/themes_colors.R")

```

# Read in the data
```{r}

#Download data from GEO
#see here for a description of different download formats: https://bioconductor.org/packages/devel/bioc/vignettes/GEOquery/inst/doc/GEOquery.html
mat <- getGEO("GSE...")
GSE... <- getGEO('GSE...',GSEMatrix=TRUE)
show(GSE...)

#Or get individual files: 
#count matrix
mat <- readMM(file = file.path(data.dir, "GSE136831_RawCounts_Sparse.mtx.gz"))

#barcodes / cells
cells <- readLines(con =  file.path(data.dir, "GSE136831_AllCells.cellBarcodes.txt.gz"))

#features
features <- read.table(file = file.path(data.dir, "GSE136831_AllCells.GeneIDs.txt.gz"), header = TRUE)

#Use biomart to get meta data 
# biomaRt was being unreliable
# mart <- useDataset(
#     "hsapiens_gene_ensembl",
#     useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = 'ensembl.org')
# )
#
# genes <- grep(pattern = "ENSG", x = features[, 2], value = TRUE)
# gene_IDs <- rbind(
#   getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","hgnc_symbol"), values = genes[1:5000], mart = mart),
#   getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","hgnc_symbol"), values = genes[5001:10000], mart = mart),
#   getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","hgnc_symbol"), values = genes[10001:length(x = genes)], mart = mart)
# )

#get meta data directly
gene_IDs <- read.csv(file = file.path(data.dir, "adams_gene_ids.csv"))


#add metadata 
rownames(x = mat) <- features[, 2]
colnames(x = mat) <- cells
meta <- read.table(file = file.path(data.dir, "GSE136831_AllCells.Samples.CellType.MetadataTable.txt.gz"), row.names = 1, head = TRUE)
mat <- as(mat, "dgCMatrix")

for (i in 1:nrow(x = gene_IDs)) {
  if (gene_IDs[i, "hgnc_symbol"] != "") {
    rn.idx <- which(x = rownames(x = mat) == gene_IDs[i, "ensembl_gene_id"])
    if (gene_IDs[i, "hgnc_symbol"] %in% rownames(x = mat)) {
      mat[gene_IDs[i, "hgnc_symbol"], ] <- mat[gene_IDs[i, "hgnc_symbol"], ] + mat[gene_IDs[i, "ensembl_gene_id"], ]
      mat <- mat[-c(rn.idx), ]
    } else {
      rownames(x = mat)[rn.idx] <- gene_IDs[i, "hgnc_symbol"]
    }
  }
}

gc()


```
