---
title: "Integration and UMAPs for epithelial libraries"
author: "Holly A. R. Giles"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
      code_folding: "hide" 
---

------------------------------------------------------------------------

This vignette explains the steps to go through from raw fastq files downloaded from the CRUK core facility, through to analysis with Seurat. It also demonstrates different packages for cell-cell communication analyses and pseudotime analsyes.

# Initial QC check

1.  Check all data files are present

What to expect (if generated by bcl2fastq)

Fastq files (for single cell RNAseq, you will get r_1, r_2, i_1 and i_2)

```         
<SLX>.<barcode>.<flowcell>.s_<lane>.r_<read>.fq.gz
```

Lost reads (you can usually ignore)

```         
<SLX>.<flowcell>.s_<lane>.lostreads.fq.gz
```

A checksum files for each sample and the lost reads.

```         
<SLX>.<barcode>.<flowcell>.s_<lane>.md5sums.txt
<SLX>.<flowcell>.s_<lane>.lostreads.md5sums.txt
```

Plus some additional outputs from bcl2fastq:

A folder ending in bcl2fastq containing stats reports and meta data

A \<SLX\>.\<flowcell\>.s\_\<lane\>.html file containing QC report

\<SLX\>.\<flowcell\>.s\_\<lane\>.bcl2fastq.zip: This is a zip archive of bcl2fastq's Stats folder and the sample sheet used to create the FASTQ files. The statistics are XML and JSON files that can be parsed to extract the numbers of reads, yield, unknown barcodes and so forth. It is created by bcl2fastq so you can refer to Illumina's documentation for details.

\<SLX\>.\<flowcell\>.s\_\<lane\>.contents.csv: A small CSV file detailing the barcoding of samples in the pool. Gives the sample name, barcode name and barcode sequence for the pool. May not be available for those sequencing methods that do not have multiplexing information in Clarity (no index, inline barcode, custom index).

2.  Run an mds5sums check

-   You can make sure that the local copy of the files match the original by running the md5sum command once the download has finished.

    Just run

    ```         
    md5sum -c *.md5sums.txt
    ```

-   3\. Take a look at the QC file

    This is a [MultiQC](https://multiqc.info/) report containing the individual report. MultiQC searches a given directory for analysis logs and compiles a HTML report. It's a general use tool, perfect for summarising the output from numerous bioinformatics tools. these include:

    -   General Statistics

    -   Demuliplexing and barcode balance reports. These give charts and numbers for the reads created from the sequencing. If there are indexing problems they'll show up here.

    -   Single cell reports. These are only present for 10x single cell lanes and are produced per sample.

    -   [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/): Assorted sequencing metrics. Where the lane is regular paired end sequencing, there will be a FastQC report for each read (FastQC is not a tool that handles paired end).

    -   [Multi Genome Alignment](https://github.com/crukci-bioinformatics/MGA): A contaminant screen for the pool.

    More details below:

-   General Statistics:

    -   This gives a quick summary for each sample:

        -   Number of reads assigned to samples - check this is high

        -   Number of lost reads - check this isn't too high

        -   \% Duplications - with RNAseq this can be a little higher, fastqc overestimates and isn't very forgiving. Some genes are really highly expressed so there's lots of them in the data

        -   %GC - make sure this fits with what it should be for the organism and that its normally distributed, no spikes

    -   FastQC:

        -   Sequence counts - in RNA seq you will have many copies of transcripts, so lots of duplicates, check none look freaky

        -   Mean quality scores - check all pass

        -   Pre-sequence quality scores - check all pass

        -   Per base seq content - bases should all be similar proportions. Don't worry too much about warnings = RNAseq libraries will often have a bias just because of how they produced - its should all be the same letter over represented though. See [here](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html)

        -   Per base N content - check all pass

        -   Sequence Length Distribution - check all pass

        -   Per Seq GC - check that they all pass, are close to the expected value for that organism and are normally distributed

        -   Seq duplication levels - they might not all pass, but check that the lines are all similar, no freaky samples - maybe expect 30% duplications. then check that any one sequence isn't too much over-represented (maybe \>1% each), otherwise a single sequence e.g an adaptor is dominating

        -   Adapter Content - check all pass

    -   Multi Genome Alignment - this checks pre-alignment what species DNA is there, and then checks what aligned. Make sure its all human that's being aligned

    -   Barcode balance - check each sample is sequenced relatively evenly, and that there aren't too many lost reads

    -   Single Cell Summary

        -   Check summary looks all good - high percentage mapped, very little mito UMIs compared to other UMIS per cell, lots of genes (not lots of the same gene)

        -   Reads aligned to referece - check not freaky

        -   Genomic Alignment - check not freaky

        -   Plots

            -   Check if you cut out cells based on UMIs / Mt reads you're not going to cut all your sample

-   

# Running cellRanger

1.  Change the file naming to be compatible with cellRanger pipelines

    1.  Use this script: csci_to_cellRanger.py

    2.  Execute: `python3 csci_to_cellRanger.py` in the folder with the fastqs

2.  Add a reference file to the folder containing your fastqs

    1.  Check you have the most update to date reference (check this is the most up to date [here](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest))

    2.  Download the reference

        `wget https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz`

    3.  Then unpack the tar file:

        `tar -xzvf refdata-gex-GRCh38-2020-A.tar.gz`

3.  Set up your cellRanger counts script

    1.  Use the template `count.sh`

    2.  Update the library names

    3.  Update the reference name

4.  Run on a screen

    1.  `screen`

    2.  ctrl + a + d to detach the screen

    3.  screen -r to resume screen

5.  

6.  

# Running Seurat to generate a UMAP

# Downstream analyses

## Differential expression analysis

## Pseudotime analyses

## Cell-Cell communication analyses
